crmsh/join_lock.py

```diff
@@ -0,0 +1,154 @@
+    SSH_EXIT_ERR = 255
+    MIN_JOIN_TIMEOUT = 120
+    WAIT_INTERVAL = 10
+
+    def __init__(self, init_node):
+        """
+        Init function
+        """
+        self.init_node = init_node
+
+    @property
+    def join_timeout(self):
+        """
+        Get join_timeout from config.core
+        """
+        try:
+            value = int(config.core.join_timeout)
+        except ValueError:
+            raise ValueError("Invalid format of core.join_timeout(should be a number)")
+        if value < self.MIN_JOIN_TIMEOUT:
+            raise ValueError("Minimum value of core.join_timeout should be {}".format(self.MIN_JOIN_TIMEOUT))
+        return value
+
+    def _run(self, cmd):
+        """
+        Run command on target node, consider specific exceptions
+        """
+        cmd_with_ssh = "ssh {} root@{} \"{}\"".format(self.SSH_OPTION, self.init_node, cmd)
+        rc, out, err = utils.get_stdout_stderr(cmd_with_ssh)
+        if rc == self.SSH_EXIT_ERR:
+            raise SSHError(err)
+        return rc, out, err
+
+    def _create_lock_dir(self):
+        """
+        Create lock directory, mkdir command was atomic
+        """
+        rc, _, _ = self._run(self.MKDIR_CMD)
+        return rc == 0
+
+    def _get_online_nodelist(self):
+        """
+        Get the online node list from init node
+        """
+        rc, out, err = self._run("crm_node -l")
+        if rc != 0 and err:
+            raise RuntimeError(err)
+        return re.findall('[0-9]+ (.*) member', out)
+
+    def _reset_timer(self):
+        """
+        Reset timer
+        """
+        self.current_time = int(time.time())
+        self.timeout = self.current_time + self.join_timeout
+
+    def _increase_timer(self):
+        """
+        Update current_time to close the timeout
+        """
+        self.current_time = int(time.time())
+
+    @property
+    def _timed_out(self):
+        """
+        Check whether reached the timeout
+        """
+        return self.current_time > self.timeout
+
+    def _lock_or_wait(self):
+        """
+        Try to claim lock on init node,
+        wait if failed to claim
+        exit if reached the join_timeout
+        """
+        warned_once = False
+        online_list = []
+        pre_online_list = []
+        expired_error_str = "Cannot continue since the lock directory exists at the init node ({}:{})".format(self.init_node, self.JOIN_LOCK_DIR)
+
+        self._reset_timer()
+        while not self._timed_out:
+
+            # Try to claim the lock
+            if self._create_lock_dir():
+                # Success
+                break
+
+            # Might lose claiming lock again, start to wait again
+            online_list = self._get_online_nodelist()
+            if pre_online_list and pre_online_list != online_list:
+                warned_once = False
+                self._reset_timer()
+                continue
+            else:
+                pre_online_list = online_list
```

Comment:
Assuming there were 3+ nodes join the init node in parallel, some nodes might lose claiming again after one of them successfully joined the cluster, since there were still at least 2 nodes try to join.
So, when we know the membership changed(by `_get_online_nodelist`), just reset timer and `warned_once` flag and continue looping